File: src/FluentReflection/Accessors/IEventAccessor.cs
using System.Reflection;

namespace FluentReflection.Accessors;

/// <summary>
/// Fluent accessor for interacting with events and their subscribers.
/// </summary>
public interface IEventAccessor
{
    /// <summary>
    /// Gets the invocation list of delegates subscribed to the event.
    /// </summary>
    /// <returns>An array of delegates representing the event subscribers, or an empty array if there are no subscribers.</returns>
    Delegate[] GetInvocationList();

    /// <summary>
    /// Subscribes the given delegate to the event.
    /// </summary>
    /// <param name="handler">The delegate to subscribe.</param>
    void Add(Delegate handler);

    /// <summary>
    /// Unsubscribes the given delegate from the event.
    /// </summary>
    /// <param name="handler">The delegate to unsubscribe.</param>
    void Remove(Delegate handler);
}

internal class EventAccessor(Type targetType, object? instance, string name) : IEventAccessor
{
    private BindingFlags Flags =>
        BindingFlags.NonPublic |
        BindingFlags.Public |
        (instance is null ? BindingFlags.Static : BindingFlags.Instance);

    public Delegate[] GetInvocationList()
    {
        FieldInfo? field = GetFieldInfo(targetType, name)
            ?? GetFieldInfo(targetType, $"<{name}>k__BackingField")
            ?? GetFieldInfo(targetType, $"_{name}");

        if (field is null)
        {
            var eventInfo = GetEventInfo(targetType, name);
            if (eventInfo is not null && eventInfo.DeclaringType is { } declaringType)
            {
                field = GetFieldInfo(declaringType, name)
                    ?? GetFieldInfo(declaringType, $"<{name}>k__BackingField")
                    ?? GetFieldInfo(declaringType, $"_{name}");
            }
        }

        if (field is null)
        {
            throw new MissingMemberException($"Event or backing field for event '{name}' not found on type '{targetType.FullName}'.");
        }

        var fieldValue = field.GetValue(instance);
        if (fieldValue is Delegate del)
        {
            return del.GetInvocationList();
        }

        return Array.Empty<Delegate>();
    }

    private FieldInfo? GetFieldInfo(Type type, string fieldName)
    {
        Type? current = type;
        while (current != null)
        {
            var field = current.GetField(fieldName, Flags);
            if (field != null) return field;
            current = current.BaseType;
        }
        return null;
    }

    public void Add(Delegate handler)
    {
        var eventInfo = GetEventInfo(targetType, name)
            ?? throw new MissingMemberException($"Event '{name}' not found on type '{targetType.FullName}'.");

        var addMethod = eventInfo.GetAddMethod(true)
            ?? throw new MissingMemberException($"Event '{name}' does not expose an add accessor.");

        addMethod.Invoke(instance, [handler]);
    }

    public void Remove(Delegate handler)
    {
        var eventInfo = GetEventInfo(targetType, name)
            ?? throw new MissingMemberException($"Event '{name}' not found on type '{targetType.FullName}'.");

        var removeMethod = eventInfo.GetRemoveMethod(true)
            ?? throw new MissingMemberException($"Event '{name}' does not expose a remove accessor.");

        removeMethod.Invoke(instance, [handler]);
    }

    private EventInfo? GetEventInfo(Type type, string eventName)
    {
        Type? current = type;
        while (current != null)
        {
            var eventInfo = current.GetEvent(eventName, Flags);
            if (eventInfo != null) return eventInfo;
            current = current.BaseType;
        }
        return null;
    }
}

File: docs/data/IEventAccessor.js
window.FR_TYPES = window.FR_TYPES || {};
window.FR_TYPES["IEventAccessor"] = {
  "name": "IEventAccessor",
  "namespace": "FluentReflection.Accessors",
  "kind": "Interface",
  "since": "1.0.0.3",
  "declaration": "public interface IEventAccessor",
  "overline": "Namespace: FluentReflection.Accessors · Assembly: FluentReflection.NET",
  "sections": [
    {
      "heading": "Remarks",
      "html": "<p>Fluent accessor for interacting with events and their subscribers. Obtained via\n       <a href=\"../IClass/index.html#Event\">IClass.Event(string)</a>. Use it to introspect which\n       delegates are subscribed to a (potentially private) event without invoking it.</p><p>The accessor locates the backing delegate field using several naming conventions, and walks\n       the base-type hierarchy to support inherited events. Note: although the public contract is\n       the interface below, the concrete implementation (<code>EventAccessor</code>) is internal;\n       access it only through the fluent API.</p>"
    }
  ],
  "members": [
    {
      "name": "GetInvocationList()",
      "kind": "Method",
      "since": "1.0.0.3",
      "href": "getinvocationlist.html",
      "return": "Delegate[]",
      "desc": "Gets the invocation list of delegates subscribed to the event.",
      "declaration": "Delegate[] GetInvocationList()",
      "overline": "IEventAccessor · Returns: Delegate[]",
      "sections": [
        {
          "heading": "Definition",
          "html": "<p>Gets the invocation list of delegates subscribed to the event.</p>"
        },
        {
          "heading": "Returns",
          "html": "<p>An array of delegates representing the event subscribers, or an empty array if there are no subscribers.</p>"
        },
        {
          "heading": "Remarks",
          "html": "<p>The backing field is located by trying, in order: the directly named field, the compiler-generated (<code>&lt;Name&gt;k__BackingField</code>), the underscore-prefixed (<code>_Name</code>) convention, then the event's declaring type.</p>"
        },
        {
          "heading": "Exceptions",
          "html": "<table><tr><th>Type</th><th>Condition</th></tr><tr><td>MissingMemberException</td><td>No event or backing field with the given name was found on the target type.</td></tr></table>"
        }
      ]
    },
    {
      "name": "Add(Delegate handler)",
      "kind": "Method",
      "since": "1.0.0.6",
      "href": "add.html",
      "return": "void",
      "desc": "Subscribes a delegate to the event.",
      "declaration": "void Add(Delegate handler)",
      "overline": "IEventAccessor · Returns: void",
      "sections": [
        {
          "heading": "Definition",
          "html": "<p>Subscribes the given delegate to the event.</p>"
        },
        {
          "heading": "Parameters",
          "html": "<p class=\"emphasized\"><b>Parameters</b></p><table><tr><th>Name</th><th>Type</th><th>Description</th></tr>\n<tr><td><code>handler</code></td><td>Delegate</td><td>The delegate to subscribe.</td></tr>\n</table>"
        },
        {
          "heading": "Remarks",
          "html": "<p>Invokes the event's add accessor via reflection, supporting private and non-public add accessors.</p>"
        },
        {
          "heading": "Exceptions",
          "html": "<table><tr><th>Type</th><th>Condition</th></tr><tr><td>MissingMemberException</td><td>No event with the given name exists on the target type, or it does not expose an add accessor.</td></tr></table>"
        }
      ]
    },
    {
      "name": "Remove(Delegate handler)",
      "kind": "Method",
      "since": "1.0.0.6",
      "href": "remove.html",
      "return": "void",
      "desc": "Unsubscribes a delegate from the event.",
      "declaration": "void Remove(Delegate handler)",
      "overline": "IEventAccessor · Returns: void",
      "sections": [
        {
          "heading": "Definition",
          "html": "<p>Unsubscribes the given delegate from the event.</p>"
        },
        {
          "heading": "Parameters",
          "html": "<p class=\"emphasized\"><b>Parameters</b></p><table><tr><th>Name</th><th>Type</th><th>Description</th></tr>\n<tr><td><code>handler</code></td><td>Delegate</td><td>The delegate to unsubscribe.</td></tr>\n</table>"
        },
        {
          "heading": "Remarks",
          "html": "<p>Invokes the event's remove accessor via reflection, supporting private and non-public remove accessors.</p>"
        },
        {
          "heading": "Exceptions",
          "html": "<table><tr><th>Type</th><th>Condition</th></tr><tr><td>MissingMemberException</td><td>No event with the given name exists on the target type, or it does not expose a remove accessor.</td></tr></table>"
        }
      ]
    }
  ],
  "properties": []
};
